##### Step 1: Install the ipython-sql and psycopg2 libraries to connect to PostgreSQL

In [1]:
!pip freeze | grep -E 'ipython-sql|psycopg2'

ipython-sql==0.4.1
psycopg2==2.9.5
psycopg2-binary==2.9.5


##### Step 2: Checking the version of PostgreSQL

In [2]:
!psql --version

psql (PostgreSQL) 12.20 (Ubuntu 12.20-0ubuntu0.20.04.1)


##### Step 3: Drop the database if exists

In [3]:
!dropdb -U student DMA_G6

dropdb: error: database removal failed: ERROR:  database "DMA_G6" does not exist


##### Step 4: Create database 

In [2]:
!createdb -U student DMA_G6

createdb: error: database creation failed: ERROR:  database "DMA_G6" already exists


##### Step 5: Load the SQL extension to run SQL queries within the notebook cells

In [3]:
%load_ext sql

##### Step 6: Connecting to our database

In [4]:
%sql postgresql://student@/DMA_G6

##### Step 7: Creating SUB - Drop if it exists and then create the new SUB table in the database. 

* The CREATE TABLE "TABLENAME" is followed by all the attributes along with their datatypes. 

In [7]:
%%sql
DROP TABLE IF EXISTS SUB CASCADE;
CREATE TABLE SUB (
    adsh VARCHAR(20) PRIMARY KEY,        -- Accession Number
    cik NUMERIC(10) NOT NULL,            -- Central Index Key (CIK)
    name VARCHAR(150) NOT NULL,          -- Name of registrant
    sic NUMERIC(4),                      -- Standard Industrial Classification (SIC)
    countryba VARCHAR(2),                -- Business address country (ISO 3166-1)
    stprba VARCHAR(2),                   -- Business address state/province (if US/CA)
    cityba VARCHAR(30),                  -- Business address city
    zipba VARCHAR(10),                   -- Business address zip code
    bas1 VARCHAR(40),                    -- Business address street line 1
    bas2 VARCHAR(40),                    -- Business address street line 2
    baph VARCHAR(20),                    -- Business address phone number
    countryma VARCHAR(2),                -- Mailing address country (ISO 3166-1)
    stprma VARCHAR(2),                   -- Mailing address state/province (if US/CA)
    cityma VARCHAR(30),                  -- Mailing address city
    zipma VARCHAR(10),                   -- Mailing address zip code
    mas1 VARCHAR(40),                    -- Mailing address street line 1
    mas2 VARCHAR(40),                    -- Mailing address street line 2
    countryinc VARCHAR(2),               -- Country of incorporation (ISO 3166-1)
    stprinc VARCHAR(2),                  -- State/province of incorporation (if US/CA)
    ein NUMERIC(10),                     -- Employee Identification Number (EIN)
    former VARCHAR(150),                 -- Most recent former name of the registrant
    changed VARCHAR(8),                  -- Date of change from former name
    afs VARCHAR(5),                      -- Filer status with the Commission
    wksi BOOLEAN NOT NULL,               -- Well Known Seasoned Issuer (WKSI)
    fye VARCHAR(4),                      -- Fiscal Year End Date (mmdd)
    form VARCHAR(20) NOT NULL,           -- Submission type
    period DATE,                         -- Balance sheet date
    fy INTEGER,                          -- Fiscal Year Focus
    fp VARCHAR(2),                       -- Fiscal Period Focus (FY, Q1, Q2, etc.)
    filed DATE NOT NULL,                 -- Date of filing with the Commission
    accepted TIMESTAMP NOT NULL,         -- Acceptance date and time
    prevrpt BOOLEAN NOT NULL,            -- Previous Report
    detail BOOLEAN NOT NULL,             -- Detail level in XBRL submission
    instance VARCHAR(255),               -- XBRL Instance Document
    nciks NUMERIC(4) NOT NULL,           -- Number of Central Index Keys (CIKs)
    aciks VARCHAR(120),                  -- Additional CIKs (space delimited)
    pubfloatusd NUMERIC(20,2),           -- Increased precision for Public float in USD
    floatdate DATE,                      -- Date public float was measured
    floataxis VARCHAR(255),              -- Nature of public float summation
    floatmems NUMERIC(2)                 -- Updated precision for floatmems to allow larger values
);


 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 8: Add comments on SUB table and all of its attributes

In [8]:
%%sql

-- Comment on SUB table
COMMENT ON TABLE SUB IS 'Submission data set; this includes one record for each XBRL submission. The set includes fields of information pertinent to the submission and the filing entity. Information is extracted from the Commission''s EDGAR system and the filings submitted to the Commission by registrants.';

-- Comment on attributes of SUB table
COMMENT ON COLUMN SUB.adsh IS 'Accession Number. The 20-character string formed from the 18-digit number assigned by the Commission to each EDGAR submission.';
COMMENT ON COLUMN SUB.cik IS 'Central Index Key (CIK). Ten-digit number assigned by the Commission to each registrant that submits filings.';
COMMENT ON COLUMN SUB.name IS 'Name of registrant. This corresponds to the name of the legal entity as recorded in EDGAR as of the filing date.';
COMMENT ON COLUMN SUB.sic IS 'Standard Industrial Classification (SIC). Four-digit code assigned by the Commission indicating the registrant''s type of business.';
COMMENT ON COLUMN SUB.countryba IS 'The ISO 3166-1 country of the registrant''s business address.';
COMMENT ON COLUMN SUB.stprba IS 'The state or province of the registrant''s business address, if field countryba is US or CA.';
COMMENT ON COLUMN SUB.cityba IS 'The city of the registrant''s business address.';
COMMENT ON COLUMN SUB.zipba IS 'The zip code of the registrant''s business address.';
COMMENT ON COLUMN SUB.bas1 IS 'The first line of the street of the registrant''s business address.';
COMMENT ON COLUMN SUB.bas2 IS 'The second line of the street of the registrant''s business address.';
COMMENT ON COLUMN SUB.baph IS 'The phone number of the registrant''s business address.';
COMMENT ON COLUMN SUB.countryma IS 'The ISO 3166-1 country of the registrant''s mailing address.';
COMMENT ON COLUMN SUB.stprma IS 'The state or province of the registrant''s mailing address, if field countryma is US or CA.';
COMMENT ON COLUMN SUB.cityma IS 'The city of the registrant''s mailing address.';
COMMENT ON COLUMN SUB.zipma IS 'The zip code of the registrant''s mailing address.';
COMMENT ON COLUMN SUB.mas1 IS 'The first line of the street of the registrant''s mailing address.';
COMMENT ON COLUMN SUB.mas2 IS 'The second line of the street of the registrant''s mailing address.';
COMMENT ON COLUMN SUB.countryinc IS 'The country of incorporation for the registrant.';
COMMENT ON COLUMN SUB.stprinc IS 'The state or province of incorporation for the registrant, if countryinc is US or CA, otherwise NULL.';
COMMENT ON COLUMN SUB.ein IS 'Employee Identification Number. A unique 9-digit identification number assigned by the IRS to business entities operating in the United States.';
COMMENT ON COLUMN SUB.former IS 'Most recent former name of the registrant, if any.';
COMMENT ON COLUMN SUB.changed IS 'Date of change from the former name, if any.';
COMMENT ON COLUMN SUB.afs IS 'Filer status with the Commission at the time of submission: 1-LAF=Large Accelerated, 2-ACC=Accelerated, 3-SRA=Smaller Reporting Accelerated, 4-NON=Non-Accelerated, 5-SML=Smaller Reporting Filer, NULL=not assigned.';
COMMENT ON COLUMN SUB.wksi IS 'Well Known Seasoned Issuer (WKSI). An issuer that meets specific Commission requirements at some point during a 60-day period preceding the date the issuer satisfies its obligation to update its shelf registration statement.';
COMMENT ON COLUMN SUB.fye IS 'Fiscal Year End Date.';
COMMENT ON COLUMN SUB.form IS 'The submission type of the registrant''s filing.';
COMMENT ON COLUMN SUB.period IS 'Balance Sheet Date.';
COMMENT ON COLUMN SUB.fy IS 'Fiscal Year Focus (as defined in EFM Ch. 6).';
COMMENT ON COLUMN SUB.fp IS 'Fiscal Period Focus within Fiscal Year, with possible values including FY, Q1, Q2, Q3, etc.';
COMMENT ON COLUMN SUB.filed IS 'The date of the registrant''s filing with the Commission.';
COMMENT ON COLUMN SUB.accepted IS 'The acceptance date and time of the registrant''s filing with the Commission.';
COMMENT ON COLUMN SUB.prevrpt IS 'Previous Report – TRUE indicates that the submission information was subsequently amended prior to the end cutoff date of the data set.';
COMMENT ON COLUMN SUB.detail IS 'TRUE indicates that the XBRL submission contains quantitative disclosures within the footnotes and schedules at the required detail level.';
COMMENT ON COLUMN SUB.instance IS 'The name of the submitted XBRL Instance Document (EX-101.INS) type data file.';
COMMENT ON COLUMN SUB.nciks IS 'Number of Central Index Keys (CIK) of registrants included in the consolidating entity''s submitted filing.';
COMMENT ON COLUMN SUB.aciks IS 'Additional CIKs of co-registrants included in a consolidating entity''s EDGAR submission, separated by spaces.';
COMMENT ON COLUMN SUB.pubfloatusd IS 'Public float, in USD, if provided in this submission.';
COMMENT ON COLUMN SUB.floatdate IS 'Date on which the public float was measured by the filer.';
COMMENT ON COLUMN SUB.floataxis IS 'If the public float value was computed by summing across several tagged values, this indicates the nature of the summation.';
COMMENT ON COLUMN SUB.floatmems IS 'If the public float was computed, the number of terms in the summation.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

##### Step 9: Creating TAG - Drop if it exists and then create the new TAG table in the database

In [9]:
%%sql
DROP TABLE IF EXISTS TAG CASCADE;
CREATE TABLE TAG (
    tag VARCHAR(256) NOT NULL,             -- Unique identifier for the tag in a taxonomy release
    version VARCHAR(20) NOT NULL,          -- Taxonomy identifier or accession number
    custom BOOLEAN NOT NULL,               -- 1 if custom (version = adsh), 0 if standard
    abstract BOOLEAN NOT NULL,             -- 1 if not representing a numeric fact
    datatype VARCHAR(20),                  -- Data type (if abstract = 1, then NULL)
    iord CHAR(1),                          -- "I" for point-in-time, "D" for duration (if abstract = 1, then NULL)
    crdr CHAR(1),                          -- "C" for credit, "D" for debit (if datatype = monetary, else NULL)
    tlabel VARCHAR(512),                   -- Label text for the tag
    doc VARCHAR(2048),                     -- Detailed definition for the tag (truncated to 2048 characters)

    PRIMARY KEY (tag, version)             -- Set a primary key on tag and version
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 10: Add comments on TAG table and all of its attributes 

In [10]:
%%sql

-- Comment on TAG table
COMMENT ON TABLE TAG IS 'Tag data set; includes defining information about each tag. Information includes tag descriptions (documentation labels), taxonomy version information, and other tag attributes.';

-- Comment on attributes of TAG table
COMMENT ON COLUMN TAG.tag IS 'The unique identifier (name) for a tag in a specific taxonomy release.';
COMMENT ON COLUMN TAG.version IS 'For a standard tag, an identifier for the taxonomy; otherwise the accession number where the tag was defined.';
COMMENT ON COLUMN TAG.custom IS '1 if the tag is custom (version=adsh), 0 if it is standard. This flag is technically redundant with the version and adsh fields.';
COMMENT ON COLUMN TAG.abstract IS '1 if the tag is not used to represent a numeric fact.';
COMMENT ON COLUMN TAG.datatype IS 'The data type for the tag (e.g., monetary). If abstract=1, then NULL.';
COMMENT ON COLUMN TAG.iord IS 'Indicates if the value is a point in time or a duration. If abstract=1, then NULL.';
COMMENT ON COLUMN TAG.crdr IS 'The tag''s natural accounting balance (debit or credit). If not defined, then NULL.';
COMMENT ON COLUMN TAG.tlabel IS 'The label text provided by the taxonomy or the filer. A tag which had neither would have a NULL value here.';
COMMENT ON COLUMN TAG.doc IS 'The detailed definition for the tag, truncated to 2048 characters. If a standard tag, then the text provided by the taxonomy, otherwise the text assigned by the filer.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

##### Step 11: Creating DIM - Drop if it exists and then create the new DIM table in the database

In [11]:
%%sql
DROP TABLE IF EXISTS DIM CASCADE;
CREATE TABLE DIM (
    dimh CHAR(64) PRIMARY KEY,           -- MD5 hash of the segments field text (64-character hex)
    segments VARCHAR(1024) ,             -- Concatenation of tag names representing axis and members
    segt BOOLEAN NOT NULL                -- TRUE if the segments field was truncated, else FALSE
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 12: Add comments on DIM table and all of its attributes

In [12]:
%%sql

-- Comment on DIM table
COMMENT ON TABLE DIM IS 'Dimension data set; used to provide more detail in numeric and non-numeric facts.';

-- Comment on attributes of DIM table
COMMENT ON COLUMN DIM.dimh IS 'MD5 hash of the segments field text, used as a primary key. Although MD5 is unsuitable for cryptographic use, it is used here merely to limit the size of the primary key.';
COMMENT ON COLUMN DIM.segments IS 'Concatenation of tag names representing the axis and members appearing in the XBRL segments. Each dimension is represented as the pair "{axis}={member};" and the axes concatenated in lexical order.';
COMMENT ON COLUMN DIM.segt IS 'TRUE if the segments field would have been longer than 1024 characters had it not been truncated, else FALSE.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.


[]

##### Step 13: Creating REN - Drop if it exists and then create the new REN table in the database

In [13]:
%%sql
DROP TABLE IF EXISTS REN CASCADE;

CREATE TABLE REN (
    adsh VARCHAR(20) NOT NULL,                 -- Accession Number (Foreign Key to SUB table)
    report INT NOT NULL,                       -- Report grouping numeric value, matching type with PRE
    rfile CHAR(1) NOT NULL,                    -- Type of interactive data file rendered (e.g., H = .htm, X = .xml)
    menucat CHAR(2),                           -- Menu categories (C=Cover, S=Statements, N=Notes, etc.)
    shortname VARCHAR(512) ,                   -- Portion of the long name used in the renderer menu
    longname VARCHAR(512) NOT NULL,            -- Space-normalized text of the XBRL link "definition" element content
    roleuri VARCHAR(255) NOT NULL,             -- XBRL "roleuri" of the role (no foreign key constraint)
    parentroleuri VARCHAR(255),                -- XBRL "roleuri" of the parent role (nullable)
    parentreport INT,                          -- Value of the report field where roleuri matches parentroleuri (nullable)
    ultparentrpt INT,                          -- Highest ancestor report reachable by following parentreport relationships (nullable)

    PRIMARY KEY (adsh, report),                -- Composite primary key on (adsh, report)

    -- Foreign Key Constraints
    FOREIGN KEY (adsh) REFERENCES SUB(adsh)             -- Reference to SUB table (adsh)
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 14: Add comments on REN table and all of its attributes

In [14]:
%%sql

-- Comment on REN table
COMMENT ON TABLE REN IS 'Rendering data set; this contains data from the rendering of the filing on the Commission website.';

-- Comment on attributes of REN table
COMMENT ON COLUMN REN.adsh IS 'Accession Number. The 20-character string formed from the 18-digit number assigned by the Commission to each EDGAR submission.';
COMMENT ON COLUMN REN.report IS 'Represents the report grouping. The numeric value refers to the "R file" as computed by the renderer and posted on the EDGAR website. Note that in some situations the numbers skip.';
COMMENT ON COLUMN REN.rfile IS 'The type of interactive data file rendered on the EDGAR website, H = .htm file, X = .xml file.';
COMMENT ON COLUMN REN.menucat IS 'If available, one of the menu categories as computed by the renderer: C=Cover, S=Statements, N=Notes, P=Policies, T=Tables, D=Details, O=Other, and U=Uncategorized.';
COMMENT ON COLUMN REN.shortname IS 'The portion of the long name used in the renderer menu.';
COMMENT ON COLUMN REN.longname IS 'The space-normalized text of the XBRL link "definition" element content.';
COMMENT ON COLUMN REN.roleuri IS 'The XBRL "roleuri" of the role.';
COMMENT ON COLUMN REN.parentroleuri IS 'The XBRL roleuri of a role for which this role has a matching shortname prefix and a higher level menu category, as computed by the renderer.';
COMMENT ON COLUMN REN.parentreport IS 'The value of the report field for the role where roleuri equals this parentroleuri.';
COMMENT ON COLUMN REN.ultparentrpt IS 'The highest ancestor report reachable by following parentreport relationships. A note (menucat = N) is its own ultimate parent.';


 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

##### Step 15: Creating PRE - Drop if it exists and then create the new PRE table in the database

In [15]:
%%sql
DROP TABLE IF EXISTS PRE CASCADE;

CREATE TABLE PRE (
    adsh VARCHAR(20) NOT NULL,                -- Accession Number
    report INT NOT NULL,                      -- Report grouping numeric value
    line INT NOT NULL,                        -- Presentation line order for a given report
    stmt CHAR(2) ,                            -- Financial statement location (e.g., CP, BS, IS)
    inpth BOOLEAN NOT NULL,                   -- Indicates if the value is presented parenthetically
    tag VARCHAR(256) NOT NULL,                -- Tag chosen by the filer for this line item
    version VARCHAR(20) NOT NULL,             -- Taxonomy identifier if the tag is standard
    prole VARCHAR(50) NOT NULL,               -- XBRL link "role" of the preferred label
    plabel VARCHAR(512),             -- Text presented on the line item (preferred label)
    negating BOOLEAN NOT NULL,                -- Indicates if the prole is treated as negating

    PRIMARY KEY (adsh, report, line),         -- Composite primary key
    FOREIGN KEY (adsh) REFERENCES SUB(adsh),
    FOREIGN KEY (adsh, report) REFERENCES REN(adsh, report),
    FOREIGN KEY (tag, version) REFERENCES TAG(tag, version)
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 16: Add comments on PRE table and all of its attributes

In [16]:
%%sql

-- Comment on PRE table
COMMENT ON TABLE PRE IS 'Presentation data set; this provides information about how the tags and numbers were presented in the primary financial statements.';

-- Comment on attributes of PRE table
COMMENT ON COLUMN PRE.adsh IS 'Accession Number. The 20-character string formed from the 18-digit number assigned by the Commission to each EDGAR submission.';
COMMENT ON COLUMN PRE.report IS 'Represents the report grouping. The numeric value refers to the "R file" as computed by the renderer and posted on the EDGAR website. Note that in some situations the numbers skip.';
COMMENT ON COLUMN PRE.line IS 'Represents the tag''s presentation line order for a given report. Together with the statement and report field, presentation location, order and grouping can be derived.';
COMMENT ON COLUMN PRE.stmt IS 'The financial statement location to which the value of the "report" field pertains. Possible values include CP = Cover Page, BS = Balance Sheet, IS = Income Statement, CF = Cash Flow, EQ = Equity, CI = Comprehensive Income, UN = Unclassifiable Statement.';
COMMENT ON COLUMN PRE.inpth IS '1 indicates that the value was presented "parenthetically" instead of in fields within the financial statements.';
COMMENT ON COLUMN PRE.tag IS 'The tag chosen by the filer for this line item.';
COMMENT ON COLUMN PRE.version IS 'The taxonomy identifier if the tag is a standard tag; otherwise, it is the adsh.';
COMMENT ON COLUMN PRE.prole IS 'The XBRL link "role" of the preferred label, using only the portion of the role URI after the last "/".';
COMMENT ON COLUMN PRE.plabel IS 'The text presented on the line item, also known as a "preferred" label.';
COMMENT ON COLUMN PRE.negating IS 'Flag to indicate whether the prole is treated as negating by the renderer.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

##### Step 17: Creating NUM - Drop if it exists and then create the new NUM table in the database

In [17]:
%%sql
DROP TABLE IF EXISTS NUM CASCADE;
CREATE TABLE NUM (
    adsh VARCHAR(20) NOT NULL,           -- Accession Number, references SUB table
    tag VARCHAR(256) NOT NULL,           -- Tag name, references TAG table
    version VARCHAR(20) NOT NULL,        -- Taxonomy version or accession number
    ddate DATE NOT NULL,                 -- End date for the data value
    qtrs NUMERIC(8) NOT NULL,            -- Number of quarters
    uom VARCHAR(20) NOT NULL,            -- Unit of measure for the value
    dimh CHAR(34) NOT NULL,              -- Dimension hash, references DIM table
    iprx NUMERIC(2) NOT NULL,            -- Positive integer to distinguish different facts
    value NUMERIC(30,4),                 -- Value rounded to four decimal places
    footnote VARCHAR(512),               -- Footnote text
    footlen NUMERIC(6,0) NOT NULL,         -- Footnote text length
    dimn NUMERIC(1) NOT NULL,            -- Number of dimensions
    coreg VARCHAR(256),                  -- Co-registrant or parent company
    durp REAL NOT NULL,                  -- Difference between fact duration and quarter duration
    datp REAL NOT NULL,                  -- Difference between reported fact date and month-end date
    dcml NUMERIC(5,0) NOT NULL,          -- Decimal precision

    PRIMARY KEY (adsh, tag, version, ddate, qtrs, uom, dimh, iprx),

    -- Foreign Key References
    FOREIGN KEY (adsh) REFERENCES SUB(adsh),
    FOREIGN KEY (tag, version) REFERENCES TAG(tag, version),
    FOREIGN KEY (dimh) REFERENCES DIM(dimh)  -- Assuming DIM table exists and has dimh as PK
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 18: Add comments on NUM table and all of its attributes

In [18]:
%%sql

-- Comment on NUM table
COMMENT ON TABLE NUM IS 'Number data set; this includes one row for each distinct amount from each submission included in the SUB data set. The Number data set includes, for every submission, all line item values as rendered by the Commission Viewer/Previewer.';

-- Comment on attributes of NUM table
COMMENT ON COLUMN NUM.adsh IS 'Accession Number. The 20-character string formed from the 18-digit number assigned by the Commission to each EDGAR submission.';
COMMENT ON COLUMN NUM.tag IS 'The unique identifier (name) for a tag in a specific taxonomy release.';
COMMENT ON COLUMN NUM.version IS 'For a standard tag, an identifier for the taxonomy; otherwise the accession number where the tag was defined.';
COMMENT ON COLUMN NUM.ddate IS 'The end date for the data value, rounded to the nearest month end.';
COMMENT ON COLUMN NUM.qtrs IS 'The count of the number of quarters represented by the data value, rounded to the nearest whole number. "0" indicates it is a point-in-time value.';
COMMENT ON COLUMN NUM.uom IS 'The unit of measure for the value.';
COMMENT ON COLUMN NUM.dimh IS 'The 32-byte hexadecimal key for the dimensional information in the DIM data set.';
COMMENT ON COLUMN NUM.iprx IS 'A positive integer to distinguish different reported facts that otherwise would have the same primary key. For most purposes, data with iprx greater than 1 are not needed.';
COMMENT ON COLUMN NUM.value IS 'The value. This is not scaled, it is as found in the Interactive Data file, but is rounded to four digits to the right of the decimal point.';
COMMENT ON COLUMN NUM.footnote IS 'The plain text of any superscripted footnotes on the value, if any, as shown on the statement page, truncated to 512 characters.';
COMMENT ON COLUMN NUM.footlen IS 'Number of bytes in the plain text of the footnote prior to truncation; zero if no footnote.';
COMMENT ON COLUMN NUM.dimn IS 'Small integer representing the number of dimensions. Note that this value is a function of the dimension segments.';
COMMENT ON COLUMN NUM.coreg IS 'If specified, indicates a specific co-registrant, the parent company, or other entity (e.g., guarantor). NULL indicates the consolidated entity.';
COMMENT ON COLUMN NUM.durp IS 'The difference between the reported fact duration and the quarter duration (qtrs), expressed as a fraction of 1.';
COMMENT ON COLUMN NUM.datp IS 'The difference between the reported fact date and the month-end rounded date (ddate), expressed as a fraction of 1.';
COMMENT ON COLUMN NUM.dcml IS 'The value of the fact "decimals" attribute, with INF represented by 32767.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

##### Step 19: Creating TXT - Drop if it exists and then create the new TXT table in the database

In [19]:
%%sql
DROP TABLE IF EXISTS TXT CASCADE;

CREATE TABLE TXT (
    adsh VARCHAR(20) NOT NULL,               -- Accession Number (references SUB table)
    tag VARCHAR(256) NOT NULL,               -- Tag name (references TAG table)
    version VARCHAR(20) NOT NULL,            -- Taxonomy version (references TAG table)
    ddate DATE NOT NULL,                     -- Date related to the fact
    qtrs NUMERIC(8) NOT NULL,                -- Number of quarters
    iprx NUMERIC(3) NOT NULL,                -- Identifier to distinguish different facts
    lang VARCHAR(5) NOT NULL,                -- ISO language code
    dcml NUMERIC(5) NOT NULL,                -- Decimal accuracy
    durp REAL NOT NULL,                      -- Duration difference
    datp REAL NOT NULL,                      -- Date difference from month-end
    dimh CHAR(64) NOT NULL,                  -- Dimension hash (references DIM table)
    dimn NUMERIC(1) NULL,                    -- Number of dimensions (nullable)
    coreg VARCHAR(256),                      -- Co-registrant (nullable)
    escaped BOOLEAN NOT NULL,                -- Flag indicating if tags were removed
    srclen NUMERIC(8) NOT NULL,              -- Length of the source text
    txtlen NUMERIC(6),                       -- Length of whitespace-normalized text (nullable)
    footnote VARCHAR(512),                   -- Footnote content (nullable)
    footlen NUMERIC(4),                      -- Length of footnote content (nullable)
    context VARCHAR(255) NOT NULL,           -- Context reference
    value TEXT ,                             -- Text value for the tag

    PRIMARY KEY (adsh, tag, version, ddate, qtrs, dimh, iprx), -- Composite primary key

    -- Foreign Key Constraints
    FOREIGN KEY (adsh) REFERENCES SUB(adsh),              -- Reference to SUB table
    FOREIGN KEY (tag, version) REFERENCES TAG(tag, version), -- Reference to TAG table
    FOREIGN KEY (dimh) REFERENCES DIM(dimh)               -- Reference to DIM table
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 20: Add comments on TXT table and all of its attributes

In [20]:
%%sql

-- Comment on TXT table
COMMENT ON TABLE TXT IS 'Text data set; this is the plain text of all the non-numeric tagged items in the submission.';

-- Comment on attributes of TXT table
COMMENT ON COLUMN TXT.adsh IS 'Accession Number. The 20-character string formed from the 18-digit number assigned by the Commission to each EDGAR submission.';
COMMENT ON COLUMN TXT.tag IS 'The unique identifier (name) for a tag in a specific taxonomy release.';
COMMENT ON COLUMN TXT.version IS 'For a standard tag, an identifier for the taxonomy; otherwise the accession number where the tag was defined.';
COMMENT ON COLUMN TXT.ddate IS 'The end date for the data value, rounded to the nearest month end.';
COMMENT ON COLUMN TXT.qtrs IS 'The count of the number of quarters represented by the data value, rounded to the nearest whole number. A point in time value is represented by 0.';
COMMENT ON COLUMN TXT.iprx IS 'A positive integer to distinguish different reported facts that otherwise would have the same primary key. For most purposes, data with iprx greater than 1 are not needed.';
COMMENT ON COLUMN TXT.lang IS 'The ISO language code of the fact content.';
COMMENT ON COLUMN TXT.dcml IS 'The value of the fact "xml:lang" attribute, en-US represented by 32767, other "en" dialects having lower values, and other languages lower still.';
COMMENT ON COLUMN TXT.durp IS 'The difference between the reported fact duration and the quarter duration (qtrs), expressed as a fraction of 1.';
COMMENT ON COLUMN TXT.datp IS 'The difference between the reported fact date and the month-end rounded date (ddate), expressed as a fraction of 1.';
COMMENT ON COLUMN TXT.dimh IS 'The 32-byte hexadecimal key for the dimensional information in the DIM data set.';
COMMENT ON COLUMN TXT.dimn IS 'Small integer representing the number of dimensions, useful for sorting. Note that this value is function of the dimension segments.';
COMMENT ON COLUMN TXT.coreg IS 'If specified, indicates a specific co-registrant, the parent company, or other entity (e.g., guarantor). NULL indicates the consolidated entity.';
COMMENT ON COLUMN TXT.escaped IS 'Flag indicating whether the value has had tags removed.';
COMMENT ON COLUMN TXT.srclen IS 'Number of bytes in the original, unprocessed value. Zero indicates a NULL value.';
COMMENT ON COLUMN TXT.txtlen IS 'The original length of the whitespace normalized value, which may have been greater than 8192.';
COMMENT ON COLUMN TXT.footnote IS 'The plain text of any superscripted footnotes on the value, as shown on the page, truncated to 512 characters, or if there is no footnote, then this field will be blank.';
COMMENT ON COLUMN TXT.footlen IS 'Number of bytes in the plain text of the footnote prior to truncation.';
COMMENT ON COLUMN TXT.context IS 'The value of the contextRef attribute in the source XBRL document, which can be used to recover the original HTML tagging if desired.';
COMMENT ON COLUMN TXT.value IS 'The value, with all whitespace normalized, that is, all sequences of line feeds, carriage returns, tabs, non-breaking spaces, and spaces having been collapsed to a single space.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

##### Step 21: Creating CAL - Drop if it exists and then create the new CAL table in the database

In [21]:
%%sql
DROP TABLE IF EXISTS CAL CASCADE;

CREATE TABLE CAL(
    adsh VARCHAR(20) NOT NULL,                 -- Accession Number, 20-character string
    grp NUMERIC(3,0) NOT NULL,                 -- Sequential number for grouping arcs in a submission
    arc NUMERIC(3,0) NOT NULL,                 -- Sequential number for arcs within a group in a submission
    negative SMALLINT NOT NULL,                -- Indicates a weight of -1 (SMALLINT to accommodate -1)
    ptag VARCHAR(256) NOT NULL,                -- Tag for the parent of the arc
    pversion VARCHAR(20) NOT NULL,             -- Version of the tag for the parent of the arc
    ctag VARCHAR(256) NOT NULL,                -- Tag for the child of the arc
    cversion VARCHAR(20) NOT NULL,             -- Version of the tag for the child of the arc

    PRIMARY KEY (adsh, grp, arc),              -- Composite primary key

    -- Foreign Key Constraints
    FOREIGN KEY (adsh) REFERENCES SUB(adsh),                -- References adsh in SUB
    FOREIGN KEY (ptag, pversion) REFERENCES TAG(tag, version), -- References tag, version in TAG
    FOREIGN KEY (ctag, cversion) REFERENCES TAG(tag, version)  -- References tag, version in TAG
);

 * postgresql://student@/DMA_G6
Done.
Done.


[]

##### Step 22: Add comments on CAL table and all of its attributes

In [22]:
%%sql

-- Comment on CAL table
COMMENT ON TABLE CAL IS 'Calculation data set; provides information to arithmetically relate tags in a filing.';

-- Comment on attributes of CAL table
COMMENT ON COLUMN CAL.adsh IS 'Accession Number. The 20-character string formed from the 18-digit number assigned by the Commission to each EDGAR submission.';
COMMENT ON COLUMN CAL.grp IS 'Sequential number for grouping arcs in a submission.';
COMMENT ON COLUMN CAL.arc IS 'Sequential number for arcs within a group in a submission.';
COMMENT ON COLUMN CAL.negative IS 'Indicates a weight of -1 (TRUE if the arc is negative), but typically +1 (FALSE).';
COMMENT ON COLUMN CAL.ptag IS 'The tag for the parent of the arc.';
COMMENT ON COLUMN CAL.pversion IS 'The version of the tag for the parent of the arc.';
COMMENT ON COLUMN CAL.ctag IS 'The tag for the child of the arc.';
COMMENT ON COLUMN CAL.cversion IS 'The version of the tag for the child of the arc.';

 * postgresql://student@/DMA_G6
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

#### Step 23: Loading the data from the website

In [23]:
!wget -O 2024_01_SEC.zip https://www.sec.gov/files/dera/data/financial-statement-notes-data-sets/2024_01_notes.zip

--2024-10-28 07:34:12--  https://www.sec.gov/files/dera/data/financial-statement-notes-data-sets/2024_01_notes.zip
Resolving www.sec.gov (www.sec.gov)... 184.25.36.101, 2600:1408:20:a8c::17b2, 2600:1408:20:aba::17b2
Connecting to www.sec.gov (www.sec.gov)|184.25.36.101|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2024-10-28 07:34:12 ERROR 403: Forbidden.



**A 403 Forbidden error is an HTTP status code that means you do not have permission to access the requested resource. This error occurs when the web server understands the request but refuses to authorize it.**

#### Step 24: Loading the zip files and unzipping them into new folders.
**Since we could not load the data from the website we are following the approach of downloading the file for each year and unziping them into new folders.**

**The files to be unzipped are named as below**
* 2024_01_notes.zip
* 2024_02_notes.zip
* 2024_03_notes.zip
* 2024_04_notes.zip

We have used the data of 4 months for our analysis

**Code Explanation**

**!unzip**: Used to unzip the zip files.

**-o**: This flag stands for "overwrite."

-o is followed by the name of the zip file that needs to be unzipped. For example, the first line of code shows "2024_01_notes.zip".

**-d**: The -d flag specifies the destination directory where the contents will be extracted.

In [24]:
!unzip -o 2024_01_notes.zip -d ./Jan24/
!unzip -o 2024_02_notes.zip -d ./Feb24/
!unzip -o 2024_03_notes.zip -d ./Mar24/
!unzip -o 2024_04_notes.zip -d ./Apr24/

Archive:  2024_01_notes.zip
  inflating: ./Jan24/sub.tsv         
  inflating: ./Jan24/tag.tsv         
  inflating: ./Jan24/dim.tsv         
  inflating: ./Jan24/ren.tsv         
  inflating: ./Jan24/cal.tsv         
  inflating: ./Jan24/pre.tsv         
  inflating: ./Jan24/num.tsv         
  inflating: ./Jan24/txt.tsv         
  inflating: ./Jan24/readme.htm      
  inflating: ./Jan24/notes-metadata.json  
Archive:  2024_02_notes.zip
  inflating: ./Feb24/sub.tsv         
  inflating: ./Feb24/tag.tsv         
  inflating: ./Feb24/dim.tsv         
  inflating: ./Feb24/ren.tsv         
  inflating: ./Feb24/cal.tsv         
  inflating: ./Feb24/pre.tsv         
  inflating: ./Feb24/num.tsv         
  inflating: ./Feb24/txt.tsv         
  inflating: ./Feb24/readme.htm      
  inflating: ./Feb24/notes-metadata.json  
Archive:  2024_03_notes.zip
  inflating: ./Mar24/sub.tsv         
  inflating: ./Mar24/tag.tsv         
  inflating: ./Mar24/dim.tsv         
  inflating: ./Mar24/ren.tsv    

#### Step 25: Deleting the zip files
* We do this to ensure that the zip files are cleared for better storage management.
* `!rm` is the code we use to remove the zip files. 

In [25]:
!rm 2024_01_notes.zip
!rm 2024_02_notes.zip
!rm 2024_03_notes.zip
!rm 2024_04_notes.zip

#### Step 26: Importing the essential libraries and increasing the field size

**Imports the csv module: This module provides functionality to work with CSV (Comma Separated Values) files**

**Imports the sys module: This module provides access to system-specific parameters and functions, one of which is sys.maxsize.**

**csv.field_size_limit(sys.maxsize) - This line sets the maximum allowed size for individual fields within a CSV file to sys.maxsize, which is the largest positive integer that the system can handle**

In [26]:
import csv
import sys
csv.field_size_limit(sys.maxsize)

131072

#### Step 27: Converting all tsv files into csv format and giving the count for all 
* We convert the files from tsv to csv for better analysis.

**Code Explanation**
* `!csvformat -d $'\t` File_To_Be_Converted.tsv > Give_File_Name.csv - gives us the csv files, where
* `-d` is the delimiter and `$'\t'` indicates that the tsv file has tab seperated values.
* `!echo` is used to count the number of lines in the file. (We run this code to ensure that our conversion has no missed values/data)

In [27]:
# For all SUB files
!csvformat -d $'\t' Jan24/sub.tsv > Jan24/sub_1.csv
!csvformat -d $'\t' Feb24/sub.tsv > Feb24/sub_2.csv
!csvformat -d $'\t' Mar24/sub.tsv > Mar24/sub_3.csv
!csvformat -d $'\t' Apr24/sub.tsv > Apr24/sub_4.csv

In [28]:
!echo "$(wc -l Jan24/sub.tsv) $(wc -l Jan24/sub_1.csv)"
!echo "$(wc -l Feb24/sub.tsv) $(wc -l Feb24/sub_2.csv)"
!echo "$(wc -l Mar24/sub.tsv) $(wc -l Mar24/sub_3.csv)"
!echo "$(wc -l Apr24/sub.tsv) $(wc -l Apr24/sub_4.csv)"

5905 Jan24/sub.tsv 5905 Jan24/sub_1.csv
9401 Feb24/sub.tsv 9401 Feb24/sub_2.csv
9033 Mar24/sub.tsv 9033 Mar24/sub_3.csv
9635 Apr24/sub.tsv 9635 Apr24/sub_4.csv


In [29]:
# For all TAG files
!csvformat -d $'\t' Jan24/tag.tsv > Jan24/tag_1.csv
!csvformat -d $'\t' Feb24/tag.tsv > Feb24/tag_2.csv
!csvformat -d $'\t' Mar24/tag.tsv > Mar24/tag_3.csv
!csvformat -d $'\t' Apr24/tag.tsv > Apr24/tag_4.csv

In [30]:
!echo "$(wc -l Jan24/tag.tsv) $(wc -l Jan24/tag_1.csv)"
!echo "$(wc -l Feb24/tag.tsv) $(wc -l Feb24/tag_2.csv)"
!echo "$(wc -l Mar24/tag.tsv) $(wc -l Mar24/tag_3.csv)"
!echo "$(wc -l Apr24/tag.tsv) $(wc -l Apr24/tag_4.csv)"

84277 Jan24/tag.tsv 84277 Jan24/tag_1.csv
534847 Feb24/tag.tsv 534847 Feb24/tag_2.csv
420503 Mar24/tag.tsv 420503 Mar24/tag_3.csv
367714 Apr24/tag.tsv 367714 Apr24/tag_4.csv


In [31]:
# For all DIM files
!csvformat -d $'\t' Jan24/dim.tsv > Jan24/dim_1.csv
!csvformat -d $'\t' Feb24/dim.tsv > Feb24/dim_2.csv
!csvformat -d $'\t' Mar24/dim.tsv > Mar24/dim_3.csv
!csvformat -d $'\t' Apr24/dim.tsv > Apr24/dim_4.csv

In [32]:
!echo "$(wc -l Jan24/dim.tsv) $(wc -l Jan24/dim_1.csv)"
!echo "$(wc -l Feb24/dim.tsv) $(wc -l Feb24/dim_2.csv)"
!echo "$(wc -l Mar24/dim.tsv) $(wc -l Mar24/dim_3.csv)"
!echo "$(wc -l Apr24/dim.tsv) $(wc -l Apr24/dim_4.csv)"

43202 Jan24/dim.tsv 43202 Jan24/dim_1.csv
430720 Feb24/dim.tsv 430720 Feb24/dim_2.csv
277938 Mar24/dim.tsv 277938 Mar24/dim_3.csv
218979 Apr24/dim.tsv 218979 Apr24/dim_4.csv


In [33]:
# For all REN files
!csvformat -d $'\t' Jan24/ren.tsv > Jan24/ren_1.csv
!csvformat -d $'\t' Feb24/ren.tsv > Feb24/ren_2.csv
!csvformat -d $'\t' Mar24/ren.tsv > Mar24/ren_3.csv
!csvformat -d $'\t' Apr24/ren.tsv > Apr24/ren_4.csv

In [34]:
!echo "$(wc -l Jan24/ren.tsv) $(wc -l Jan24/ren_1.csv)"
!echo "$(wc -l Feb24/ren.tsv) $(wc -l Feb24/ren_2.csv)"
!echo "$(wc -l Mar24/ren.tsv) $(wc -l Mar24/ren_3.csv)"
!echo "$(wc -l Apr24/ren.tsv) $(wc -l Apr24/ren_4.csv)"

39110 Jan24/ren.tsv 39110 Jan24/ren_1.csv
275354 Feb24/ren.tsv 275354 Feb24/ren_2.csv
205022 Mar24/ren.tsv 205022 Mar24/ren_3.csv
162200 Apr24/ren.tsv 162200 Apr24/ren_4.csv


In [35]:
# For all PRE files
!csvformat -d $'\t' Jan24/pre.tsv > Jan24/pre_1.csv
!csvformat -d $'\t' Feb24/pre.tsv > Feb24/pre_2.csv
!csvformat -d $'\t' Mar24/pre.tsv > Mar24/pre_3.csv
!csvformat -d $'\t' Apr24/pre.tsv > Apr24/pre_4.csv

In [36]:
!echo "$(wc -l Jan24/pre.tsv) $(wc -l Jan24/pre_1.csv)"
!echo "$(wc -l Feb24/pre.tsv) $(wc -l Feb24/pre_2.csv)"
!echo "$(wc -l Mar24/pre.tsv) $(wc -l Mar24/pre_3.csv)"
!echo "$(wc -l Apr24/pre.tsv) $(wc -l Apr24/pre_4.csv)"

511414 Jan24/pre.tsv 511414 Jan24/pre_1.csv
3349317 Feb24/pre.tsv 3349317 Feb24/pre_2.csv
2414129 Mar24/pre.tsv 2414129 Mar24/pre_3.csv
1892638 Apr24/pre.tsv 1892638 Apr24/pre_4.csv


In [37]:
# For all NUM files
!csvformat -d $'\t' Jan24/num.tsv > Jan24/num_1.csv
!csvformat -d $'\t' Feb24/num.tsv > Feb24/num_2.csv
!csvformat -d $'\t' Mar24/num.tsv > Mar24/num_3.csv
!csvformat -d $'\t' Apr24/num.tsv > Apr24/num_4.csv

In [38]:
!echo "$(wc -l Jan24/num.tsv) $(wc -l Jan24/num_1.csv)"
!echo "$(wc -l Feb24/num.tsv) $(wc -l Feb24/num_2.csv)"
!echo "$(wc -l Mar24/num.tsv) $(wc -l Mar24/num_3.csv)"
!echo "$(wc -l Apr24/num.tsv) $(wc -l Apr24/num_4.csv)"

598844 Jan24/num.tsv 598844 Jan24/num_1.csv
5206930 Feb24/num.tsv 5206930 Feb24/num_2.csv
3291027 Mar24/num.tsv 3291027 Mar24/num_3.csv
2517965 Apr24/num.tsv 2517965 Apr24/num_4.csv


In [39]:
# For all TXT files
!csvformat -d $'\t' Jan24/txt.tsv > Jan24/txt_1.csv
!csvformat -d $'\t' Feb24/txt.tsv > Feb24/txt_2.csv
!csvformat -d $'\t' Mar24/txt.tsv > Mar24/txt_3.csv
!csvformat -d $'\t' Apr24/txt.tsv > Apr24/txt_4.csv

FieldSizeLimitError: CSV contains a field longer than the maximum length of 131072 characters on line 16247. Try raising the maximum with the field_size_limit parameter, or try setting quoting=csv.QUOTE_NONE.
FieldSizeLimitError: CSV contains a field longer than the maximum length of 131072 characters on line 28035. Try raising the maximum with the field_size_limit parameter, or try setting quoting=csv.QUOTE_NONE.
FieldSizeLimitError: CSV contains a field longer than the maximum length of 131072 characters on line 24385. Try raising the maximum with the field_size_limit parameter, or try setting quoting=csv.QUOTE_NONE.
FieldSizeLimitError: CSV contains a field longer than the maximum length of 131072 characters on line 11186. Try raising the maximum with the field_size_limit parameter, or try setting quoting=csv.QUOTE_NONE.


The **FieldSizeLimitError** indicates that a field in the CSV file exceeds the maximum allowed character limit, hence we tackle the issue by following the below code which gives out the desired output.

In [40]:
#Installing the csvtool package
!sudo apt-get install csvtool

Reading package lists... Done
Building dependency tree       
Reading state information... Done
csvtool is already the newest version (2.4-1build1).
0 upgraded, 0 newly installed, 0 to remove and 79 not upgraded.


**The following code is used to convert all .tsv files into .csv files**

**-t TAB: Specifies that the input file is tab-separated.**

**-u COMMA: Sets the output delimiter to a comma, making it a CSV format.**

**cat: The cat command in csvtool reads and processes the input file.**

In [41]:
!csvtool -t TAB -u COMMA cat Jan24/txt.tsv > Jan24/txt_1.csv
!csvtool -t TAB -u COMMA cat Feb24/txt.tsv > Feb24/txt_2.csv
!csvtool -t TAB -u COMMA cat Mar24/txt.tsv > Mar24/txt_3.csv
!csvtool -t TAB -u COMMA cat Apr24/txt.tsv > Apr24/txt_4.csv

**!echo prints out the word count for the .tsv and .csv files.**

In [42]:
!echo "$(wc -l Jan24/txt.tsv) $(wc -l Jan24/txt_1.csv)"
!echo "$(wc -l Feb24/txt.tsv) $(wc -l Feb24/txt_2.csv)"
!echo "$(wc -l Mar24/txt.tsv) $(wc -l Mar24/txt_3.csv)"
!echo "$(wc -l Apr24/txt.tsv) $(wc -l Apr24/txt_4.csv)"

204722 Jan24/txt.tsv 204722 Jan24/txt_1.csv
678421 Feb24/txt.tsv 678421 Feb24/txt_2.csv
555617 Mar24/txt.tsv 555617 Mar24/txt_3.csv
485145 Apr24/txt.tsv 485145 Apr24/txt_4.csv


In [43]:
# For all CAL files
!csvformat -d $'\t' Jan24/cal.tsv > Jan24/cal_1.csv
!csvformat -d $'\t' Feb24/cal.tsv > Feb24/cal_2.csv
!csvformat -d $'\t' Mar24/cal.tsv > Mar24/cal_3.csv
!csvformat -d $'\t' Apr24/cal.tsv > Apr24/cal_4.csv

In [44]:
!echo "$(wc -l Jan24/cal.tsv) $(wc -l Jan24/cal_1.csv)"
!echo "$(wc -l Feb24/cal.tsv) $(wc -l Feb24/cal_2.csv)"
!echo "$(wc -l Mar24/cal.tsv) $(wc -l Mar24/cal_3.csv)"
!echo "$(wc -l Apr24/cal.tsv) $(wc -l Apr24/cal_4.csv)"

52253 Jan24/cal.tsv 52253 Jan24/cal_1.csv
490774 Feb24/cal.tsv 490774 Feb24/cal_2.csv
335250 Mar24/cal.tsv 335250 Mar24/cal_3.csv
244051 Apr24/cal.tsv 244051 Apr24/cal_4.csv


#### Special Note: 
**The process from here is very extensive and can take sometime for the Kernel to process based on system speed. Please wait patiently and ensure to check the Kernal status time to time.**

#### Step 28: Stacking all 4 months csv files into a single file for each entity.

**Code Explanation**
* `!csvstack` is a command used to stack one file on another 'followed by all the .csv files that need to be stacked' > 'name of stacked file'
* Here we stack the converted .csv files of all 4 months into one final .csv file

In [45]:
#For SUB
!csvstack Jan24/sub_1.csv Feb24/sub_2.csv Mar24/sub_3.csv Apr24/sub_4.csv > stacked_sub.csv

In [46]:
#For TAG
!csvstack Jan24/tag_1.csv Feb24/tag_2.csv Mar24/tag_3.csv Apr24/tag_4.csv > stacked_tag.csv

In [47]:
#For DIM
!csvstack Jan24/dim_1.csv Feb24/dim_2.csv Mar24/dim_3.csv Apr24/dim_4.csv > stacked_dim.csv

In [48]:
#For REN
!csvstack Jan24/ren_1.csv Feb24/ren_2.csv Mar24/ren_3.csv Apr24/ren_4.csv > stacked_ren.csv

In [49]:
#For PRE
!csvstack Jan24/pre_1.csv Feb24/pre_2.csv Mar24/pre_3.csv Apr24/pre_4.csv > stacked_pre.csv

In [50]:
#For NUM
!csvstack Jan24/num_1.csv Feb24/num_2.csv Mar24/num_3.csv Apr24/num_4.csv > stacked_num.csv

#### For TXT

**Code Explanation**
* Here we use **!head** to load the header onto the stacked_txt.csv followed by **!tail -n +2** This command outputs all lines from the file, starting from the second line (i.e skips the header)
* We use this approach for Txt because of the max file size limit constraint

In [51]:
!head -n 1 Jan24/txt_1.csv > stacked_txt.csv

In [52]:
!tail -n +2 Jan24/txt_1.csv >> stacked_txt.csv
!tail -n +2 Feb24/txt_2.csv >> stacked_txt.csv
!tail -n +2 Mar24/txt_3.csv >> stacked_txt.csv
!tail -n +2 Apr24/txt_4.csv >> stacked_txt.csv

In [53]:
#For CAL
!csvstack Jan24/cal_1.csv Feb24/cal_2.csv Mar24/cal_3.csv Apr24/cal_4.csv > stacked_cal.csv

#### Step 29: Loading data on SUB table in the database

**Code Explanation**
* The **COPY** command is used to load all the data onto the database that was initially stacked onto a single file
* **Rows affected** shows that the data has been successfully loaded onto the table in our DB

In [54]:
%%sql
COPY sub FROM '/home/ubuntu/notebooks/stacked_sub.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
33970 rows affected.


[]

#### Step 30: Loading data on TAG table in the database

* We first clean the TAG data to remove duplicates as it might throw an error at the time of loading the data

**Code Explanation**
* `head -n 1`: This retrieves the first line (header) from the file mentioned
* `> clean_tag.csv`: This is used to redirect the data into a new file
* `tail -n +2`: This command retrieves all lines from the file, starting from the second line. 
* `sort -t, -k1,1 -k2,2`: Sorts the lines based on the first and second columns (-k1,1 -k2,2). 

In [55]:
!head -n 1 stacked_tag.csv > clean_tag.csv
!tail -n +2 stacked_tag.csv | sort -t, -k1,1 -k2,2 | uniq >> clean_tag.csv

In [56]:
%%sql
COPY tag FROM '/home/ubuntu/notebooks/clean_tag.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
1361021 rows affected.


[]

#### Step 31: Loading data on DIM table in the database

* We run the same command as before, just changing the variables of sorting. We do this based on the unique column according to the given metadata. In this case it is column 1 `-k1,1`

In [57]:
!head -n 1 stacked_dim.csv > clean_dim.csv
!tail -n +2 stacked_dim.csv | sort -t, -k1,1 | uniq >> clean_dim.csv

In [58]:
%%sql
COPY dim FROM '/home/ubuntu/notebooks/clean_dim.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
855682 rows affected.


[]

#### Step 32: Loading data on REN table in the database

* We run the same command as before, just changing the variables of sorting. We do this based on the unique columns according to the given metadata. In this case it is column 1 `-k1,1 -k2,2`

In [59]:
!head -n 1 stacked_ren.csv > clean_ren.csv
!tail -n +2 stacked_ren.csv | sort -t, -k1,1 -k2,2 | uniq >> clean_ren.csv

In [60]:
%%sql
COPY ren FROM '/home/ubuntu/notebooks/clean_ren.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
681682 rows affected.


[]

#### Step 33: Loading data on PRE table in the database

* We run the same command as before, just changing the variables of sorting. We do this based on the unique column according to the given metadata. In this case it is column 1 `-k1,1 -k2,2 -k3,3`

**Code Explanation**
* Given the massive size of PRE data, in order to make sure that the Kernel performs best, without any issue we breakdown the files with a line limit of **500,000** and make multiple files of it and upload it to the DB one at a time. 
* `mkdir -p pre`: Creates a directory named pre
* `split -l 500000`: Splits the file into parts, with each part containing 500,000 lines.
* `--additional-suffix=.csv`: Adds the suffix .csv to each output file to ensure they are saved in CSV format.
* `pre/pre_cleaned_part_`: Specifies the prefix for the output files, which will be saved in the pre directory with names like pre_cleaned_part_aa.csv, pre_cleaned_part_ab.csv, and so on.

In [61]:
!head -n 1 stacked_pre.csv > clean_pre.csv
!tail -n +2 stacked_pre.csv | sort -t, -k1,1 -k2,2 -k3,3 | uniq >> clean_pre.csv

In [62]:
!mkdir -p pre
!split -l 500000 --additional-suffix=.csv clean_pre.csv pre/pre_cleaned_part_

In [63]:
%%sql

-- For the PRE table
COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_aa.csv'
CSV
HEADER;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ab.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ac.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ad.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ae.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_af.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ag.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ah.csv'
CSV;


 * postgresql://student@/DMA_G6
499999 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.


[]

In [64]:
%%sql
COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ai.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [65]:
%%sql

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_aj.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ak.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_al.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_am.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_an.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ao.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_ap.csv'
CSV;

COPY PRE FROM '/home/ubuntu/notebooks/pre/pre_cleaned_part_aq.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
500000 rows affected.
167495 rows affected.


[]

#### Step 34: Loading data on NUM table in the database

* We run the same command as before, just changing the variables of sorting. We do this based on the unique column according to the given metadata. In this case it is column 1 `-k1,1 -k2,2 -k3,3 -k4,4 -k5,5 -k6,6 -k7,7 -k8,8`

**Code Explanation**
* Given the massive size of PRE data, in order to make sure that the Kernel performs best, without any issue we breakdown the files with a line limit of **500,000** and make multiple files of it and upload it to the DB one at a time. 
* `mkdir -p num`: Creates a directory named num
* `split -l 500000`: Splits the file into parts, with each part containing 500,000 lines.
* `--additional-suffix=.csv`: Adds the suffix .csv to each output file to ensure they are saved in CSV format.
* `num/num_cleaned_part_`: Specifies the prefix for the output files, which will be saved in the pre directory with names like num_cleaned_part_aa.csv, num_cleaned_part_ab.csv, and so on.

In [66]:
!head -n 1 stacked_num.csv > clean_num.csv
!tail -n +2 stacked_num.csv | sort -t, -k1,1 -k2,2 -k3,3 -k4,4 -k5,5 -k6,6 -k7,7 -k8,8 | uniq >> clean_num.csv

In [67]:
!mkdir -p num
!split -l 500000 --additional-suffix=.csv clean_num.csv num/num_cleaned_part_

In [68]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_aa.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
499999 rows affected.


[]

In [5]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ab.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [6]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ac.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [7]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ad.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [8]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ae.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [9]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_af.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [10]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ag.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [11]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ah.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [12]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ai.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [13]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_aj.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [14]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ak.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [15]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_al.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [16]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_am.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [17]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_an.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [18]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ao.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [19]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ap.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [20]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_aq.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [21]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ar.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [22]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_as.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [23]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_at.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [24]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_au.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [25]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_av.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [26]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_aw.csv'
CSV;

 * postgresql://student@/DMA_G6
500000 rows affected.


[]

In [27]:
%%sql
COPY NUM FROM '/home/ubuntu/notebooks/num/num_cleaned_part_ax.csv'
CSV;

 * postgresql://student@/DMA_G6
114763 rows affected.


[]

#### Step 35: Loading data on TXT table in the database

* We run the same command as before, just changing the variables of sorting. We do this based on the unique column according to the given metadata. In this case it is column 1 `-k1,1 -k2,2 -k3,3 -k4,4 -k5,5 -k6,6 -k11,11`

In [28]:
!head -n 1 stacked_txt.csv > clean_txt.csv
!tail -n +2 stacked_txt.csv | sort -t, -k1,1 -k2,2 -k3,3 -k4,4 -k5,5 -k6,6 -k11,11  | uniq >> clean_txt.csv

In [5]:
%%sql
COPY txt FROM '/home/ubuntu/notebooks/clean_txt.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
(psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "txt_pkey"
DETAIL:  Key (adsh, tag, version, ddate, qtrs, dimh, iprx)=(0000002178-24-000004, AmendmentFlag, dei/2023, 2024-02-29, 0, 0x00000000                                                      , 0) already exists.
CONTEXT:  COPY txt, line 2

[SQL: COPY txt FROM '/home/ubuntu/notebooks/clean_txt.csv'
CSV
HEADER;]
(Background on this error at: https://sqlalche.me/e/14/gkpj)


**When you try to load the data onto the table that already consists of the same data, it throws the above error. We tried to reload the data just to ensure that we didn't miss out on any, hence this error. Rest assured the data has been loaded fully**

#### Step 36: Loading data on CAL table in the database

* We run the same command as before, just changing the variables of sorting. We do this based on the unique column according to the given metadata. In this case it is column 1 `-k1,1 -k2,2 -k3,3`

In [6]:
!head -n 1 stacked_cal.csv > clean_cal.csv
!tail -n +2 stacked_cal.csv | sort -t, -k1,1 -k2,2 -k3,3 | uniq >> clean_cal.csv

In [7]:
%%sql
COPY cal FROM '/home/ubuntu/notebooks/clean_cal.csv'
CSV
HEADER;

 * postgresql://student@/DMA_G6
1122324 rows affected.


[]

#### Step 37: Count of all clean files followed by the count of data in the table
* We run the word count of for all the final .csv files before loading and then compare it to the word count of the data on the DB. 
* We observe that both of them match, making it clear that our data loading has been successful without any data missed out during the process.

**Code Explanation**
* `wc -l`: The wc (word count) command with the -l option counts only the lines in the specified file.
* `SELECT COUNT(*) FROM`: The COUNT(*) function returns the total number of rows in the specified table, including rows with NULL values.

In [5]:
!wc -l stacked_sub.csv
!wc -l clean_dim.csv
!wc -l clean_txt.csv
!wc -l clean_pre.csv
!wc -l clean_cal.csv
!wc -l clean_tag.csv
!wc -l clean_num.csv

33971 stacked_sub.csv
855683 clean_dim.csv
1923902 clean_txt.csv
8167495 clean_pre.csv
1122325 clean_cal.csv
1361022 clean_tag.csv
11614763 clean_num.csv


In [9]:
%%sql
select count(*) from sub;

 * postgresql://student@/DMA_G6
1 rows affected.


[(33970,)]

In [10]:
%%sql
select count(*) from dim;


 * postgresql://student@/DMA_G6
1 rows affected.


[(855682,)]

In [11]:
%%sql
select count(*) from txt;


 * postgresql://student@/DMA_G6
1 rows affected.


[(1923901,)]

In [12]:
%%sql
select count(*) from pre;


 * postgresql://student@/DMA_G6
1 rows affected.


[(8167494,)]

In [13]:
%%sql
select count(*) from cal;


 * postgresql://student@/DMA_G6
1 rows affected.


[(1122324,)]

In [14]:
%%sql
select count(*) from tag;


 * postgresql://student@/DMA_G6
1 rows affected.


[(1361021,)]

In [15]:
%%sql
select count(*) from num;

 * postgresql://student@/DMA_G6
1 rows affected.


[(11614762,)]

## Part III: DATA ANALYSIS
* Objective: Run SQL queries to answer important questions related to the data.

**Question 1: Retrieve the submission_id, company_name, and submission_date, and full
mailing address (street, city, zip and country) for all submissions made by
companies in the “Healthcare” sector in March 2024.**

**Based on the question, we need to first figure out all the Healthcare sectors in the SEC Website**
https://www.sec.gov/search-filings/standard-industrial-classification-sic-code-list

After thorough analysis, we understand that these are the list of SIC Codes that come under the healthcare sector. The Standard Industrial Classification (SIC) codes for the healthcare sector primarily include codes ranging from **8000 to 8099**

**8000 to 8099** : Are the direct healthcare relationships. 

**2833, 2834, 2835, 2836, 3841, 3842, 3845, 5047, 6324**: Are indirect healthcare sectors in pharmaceuticals, medical instruments, and insurance plans. 

* SIC 2833 - 2836: This range includes pharmaceutical products.
* 2833: Medicinal Chemicals and Botanical Products
* 2834: Pharmaceutical Preparations
* 2835: In Vitro and In Vivo Diagnostic Substances
* 2836: Biological Products, Except Diagnostic Substances
* SIC 3841, 3842, 3845: These cover medical instruments and equipment:
* 3841: Surgical and Medical Instruments and Apparatus
* 3842: Orthopedic, Prosthetic, and Surgical Appliances and Supplies
* 3845: Electromedical and Electrotherapeutic Apparatus
* SIC 5047: Medical, Dental, and Hospital Equipment and Supplies, which includes the wholesale distribution of healthcare equipment.
* SIC 6324: Hospital and Medical Service Plans, relating to health insurance services.


* SIC 8011: Offices and Clinics of Doctors of Medicine
* SIC 8021: Offices and Clinics of Dentists
* SIC 8031: Offices and Clinics of Doctors of Osteopathy
* SIC 8041: Offices and Clinics of Chiropractors
* SIC 8042: Offices and Clinics of Optometrists
* SIC 8043: Offices and Clinics of Podiatrists
* SIC 8049: Offices and Clinics of Health Practitioners, Not Elsewhere Classified
* SIC 8051: Skilled Nursing Care Facilities
* SIC 8052: Intermediate Care Facilities
* SIC 8059: Nursing and Personal Care Facilities, Not Elsewhere Classified
* SIC 8062: General Medical and Surgical Hospitals
* SIC 8063: Psychiatric Hospitals
* SIC 8069: Specialty Hospitals, Except Psychiatric
* SIC 8071: Medical Laboratories
* SIC 8082: Home Health Care Services
* SIC 8093: Specialty Outpatient Facilities, Not Elsewhere Classified
* SIC 8099: Health and Allied Services, Not Elsewhere Classified




**Code Explanation**

* `adsh AS submission_id`: Fetches the unique submission identifier, labeled as submission_id.
* `name AS company_name`: Fetches the name of the company, labeled as company_name.
* `filed AS submission_date`: Fetches the date the submission was filed, labeled as submission_date.
* `mas1 AS mailing_street, cityma AS mailing_city, zipma AS mailing_zip, countryma AS mailing_country`: Fetches the company’s mailing address details, labeled for easy identification.

In [16]:
%%sql
SELECT 
    adsh AS submission_id,
    name AS company_name,
    filed AS submission_date,
    mas1 AS mailing_street,
    cityma AS mailing_city,
    zipma AS mailing_zip,
    countryma AS mailing_country
FROM 
    SUB
WHERE 
    sic IN (2833, 2834, 2835, 2836, 3841, 3842, 3845, 5047, 6324, 8000, 8011, 8062, 8071, 8082, 8090)
    AND filed BETWEEN '2024-03-01' AND '2024-03-31'
    limit 20;

 * postgresql://student@/DMA_G6
20 rows affected.


[('0000742112-24-000007', 'INVACARE HOLDINGS CORP', datetime.date(2024, 3, 1), 'ONE INVACARE WAY', 'ELYRIA', '44035', 'US'),
 ('0000885978-24-000007', 'U S PHYSICAL THERAPY INC /NV', datetime.date(2024, 3, 1), '1300 WEST SAM HOUSTON PARKWAY', 'HOUSTON', '77043', 'US'),
 ('0000950170-24-023509', 'BIODESIX INC', datetime.date(2024, 3, 1), '2970 WILDERNESS PLACE', 'BOULDER', '80301', 'US'),
 ('0000950170-24-023513', 'BIODESIX INC', datetime.date(2024, 3, 1), '2970 WILDERNESS PLACE', 'BOULDER', '80301', 'US'),
 ('0000950170-24-023543', 'MYOMO, INC.', datetime.date(2024, 3, 1), '137 PORTLAND STREET', 'BOSTON', '02114', 'US'),
 ('0000950170-24-023820', 'INOGEN INC', datetime.date(2024, 3, 1), '859 WARD DRIVE', 'GOLETA', '93111', 'US'),
 ('0000950170-24-023838', 'VIKING THERAPEUTICS, INC.', datetime.date(2024, 3, 1), '9920 PACIFIC HEIGHTS BLVD, SUITE 350', 'SAN DIEGO', '92121', 'US'),
 ('0000950170-24-023918', 'LYRA THERAPEUTICS, INC.', datetime.date(2024, 3, 1), '480 ARSENAL WAY', 'WATERTOWN', '02472', 'US'),
 ('0000950170-24-024008', '89BIO, INC.', datetime.date(2024, 3, 1), '142 SANSOME STREET', 'SAN FRANCISCO', '94104', 'US'),
 ('0001104659-24-029624', 'EAGLE PHARMACEUTICALS, INC.', datetime.date(2024, 3, 1), '50 TICE BOULEVARD, SUITE 315', 'WOODCLIFF LAKE', '07677', 'US'),
 ('0001104659-24-029636', 'PRECISION BIOSCIENCES INC', datetime.date(2024, 3, 1), '302 EAST PETTIGREW STREET', 'DURHAM', '27701', 'US'),
 ('0001104659-24-029638', 'MARKER THERAPEUTICS, INC.', datetime.date(2024, 3, 1), '4551 KENNEDY COMMERCE DR.', 'HOUSTON', '77032', 'US'),
 ('0001104659-24-029960', 'CLEARPOINT NEURO, INC.', datetime.date(2024, 3, 1), '120 S. SIERRA AVENUE', 'SOLANA BEACH', '92075', 'US'),
 ('0001104659-24-029961', 'DARIOHEALTH CORP.', datetime.date(2024, 3, 1), '322 W 57TH ST. #33B', 'NEW YORK', '10019', 'US'),
 ('0001104659-24-029969', 'CELLDEX THERAPEUTICS, INC.', datetime.date(2024, 3, 1), '53 FRONTAGE ROAD', 'HAMPTON', '08827', 'US'),
 ('0001104659-24-030032', 'GREENBROOK TMS INC.', datetime.date(2024, 3, 1), '890 YONGE STREET, 7TH FLOOR', 'TORONTO', 'M4W 3P4', 'CA'),
 ('0001104659-24-030064', 'JAGUAR HEALTH, INC.', datetime.date(2024, 3, 1), '200 PINE STREET SUITE 400', 'SAN FRANCISCO', '94104', 'US'),
 ('0001140361-24-010584', 'KARUNA THERAPEUTICS, INC.', datetime.date(2024, 3, 1), '99 HIGH STREET', 'BOSTON', '02110', 'US'),
 ('0001140361-24-010648', 'KINNATE BIOPHARMA INC.', datetime.date(2024, 3, 1), '11975 EL CAMINO REAL, STE 101', 'SAN DIEGO', '92130', 'US'),
 ('0001145197-24-000013', 'INSULET CORP', datetime.date(2024, 3, 1), '100 NAGOG PARK', 'ACTON', '01720', 'US')]

**Question 2: On a monthly basis, calculate and display the month, the total number of
submissions and the average number of submission for the first four month
of 2024**

**Code Explanation**
* `TO_CHAR(filed, 'Month') AS month`: Converts the filed date into a month name, and aliases it as month.
* `COUNT(*) AS total_submissions`: Counts the total rows for each month, providing the number of submissions, and aliases it as total_submissions.
* `AVG(COUNT(*)) OVER () AS average_submissions`: Calculates the average number of submissions across all months in the result set. The OVER () clause means it calculates a global average over all grouped rows, aliasing it as average_submissions.
* `FROM SUB`: Specifies SUB as the table containing the data.
* `WHERE filed BETWEEN '2024-01-01' AND '2024-04-30'`: Filters rows to include only records where the filed date is between January 1, 2024, and April 30, 2024.
* `GROUP BY TO_CHAR(filed, 'Month')`: Groups the records by month (from the filed column, formatted as month names) so that COUNT(*) counts submissions per month.
* `ORDER BY MIN(filed)`: Orders the grouped results by the earliest date within each month group. This keeps the months in chronological order (January through April).

In [17]:
%%sql
SELECT 
    TO_CHAR(filed, 'Month') AS month,
    COUNT(*) AS total_submissions,
    AVG(COUNT(*)) OVER () AS average_submissions
FROM 
    SUB
WHERE 
    filed BETWEEN '2024-01-01' AND '2024-04-30'
GROUP BY 
    TO_CHAR(filed, 'Month')
ORDER BY 
    MIN(filed);


 * postgresql://student@/DMA_G6
4 rows affected.


[('January  ', 5904, Decimal('8492.5000000000000000')),
 ('February ', 9400, Decimal('8492.5000000000000000')),
 ('March    ', 9032, Decimal('8492.5000000000000000')),
 ('April    ', 9634, Decimal('8492.5000000000000000'))]

**Question 3: Retrieve the List of all tag_id, tag_name, and the total number of companies
used that tag for all tags used in submissions by companies in the “Finance”
sector in 2024**

**Upon thorough analysis we understand that these are the SIC codes related to Finance sector. Ranging between 6000-6799** 

* SIC 6000-6099: Depository Institutions (e.g., Commercial Banks, Savings Institutions)
* SIC 6100-6199: Nondepository Credit Institutions (e.g., Credit Unions, Mortgage Bankers and Brokers)
* SIC 6200-6299: Security and Commodity Brokers, Dealers, Exchanges, and Services
* SIC 6300-6399: Insurance Carriers
* SIC 6400-6499: Insurance Agents, Brokers, and Service
* SIC 6500-6599: Real Estate
* SIC 6700-6799: Holding and Other Investment Offices

**Code Explanation**

* `t.tag AS tag_id`: retrieves the unique tag identifier from the tag table.
* `t.tlabel AS tag_name` retrieves the label or description of each tag.
* `COUNT(DISTINCT s.cik) AS total_companies`: counts the distinct companies (using cik) associated with each tag in the Finance sector.
* `sub s`: Refers to submission records. The alias s is used for easier reference.
* `num n ON s.adsh = n.adsh`: Joins sub and num tables using adsh, the unique accession number linking submissions to numeric data.
* `tag t ON n.tag = t.tag`: Joins num and tag on tag, allowing access to tag descriptions and IDs.
* `s.sic BETWEEN 6000 AND 6799`: Filters for Finance sector companies based on SIC codes.
* `s.filed BETWEEN '2024-01-01' AGrouping and Ordering`:
* `GROUP BY t.tag, t.tlabel`: Groups results by tag and tag label, enabling aggregation by COUNT.
* `ORDER BY total_companies DESC`: Orders results by the count of companies using each tag, in descending order. 

In [5]:
%%sql
SELECT 
    t.tag AS tag_id,
    t.tlabel AS tag_name,
    COUNT(DISTINCT s.cik) AS total_companies
FROM 
    sub s
JOIN 
    num n ON s.adsh = n.adsh
JOIN 
    tag t ON n.tag = t.tag
WHERE 
    s.sic BETWEEN 6000 AND 6799                          -- Covers all Finance SIC codes
    AND s.filed BETWEEN '2024-01-01' AND '2024-12-31'    -- Only submissions in 2024
GROUP BY 
    t.tag, t.tlabel
ORDER BY 
    total_companies DESC
LIMIT 10;  --Adjust as needed

 * postgresql://student@/DMA_G6
10 rows affected.


[('Assets', 'Assets', 1476),
 ('EntityCommonStockSharesOutstanding', 'Entity Common Stock, Shares Outstanding', 1456),
 ('Liabilities', 'Liabilities', 1443),
 ('NetCashProvidedByUsedInOperatingActivities', 'Net Cash Provided by (Used in) Operating Activities', 1374),
 ('LiabilitiesAndStockholdersEquity', 'Liabilities and Equity', 1369),
 ('NetCashProvidedByUsedInFinancingActivities', 'Net Cash Provided by (Used in) Financing Activities', 1363),
 ('NetIncomeLoss', 'Net Income (Loss) Attributable to Parent', 1345),
 ('NetIncomeLoss', 'Net loss', 1345),
 ('NetIncomeLoss', '[Net income (loss) for the year]', 1345),
 ('NetIncomeLoss', '[Net income 1]', 1345)]

**Question 4: Retrieve the top 5 companies with the highest total assets in 2024, including
their company_name, sector, and the number of unique tags used in their
submissions.**

**Code Explanation** 
* `s.name AS company_name`: Retrieves the company’s name.
* `s.sic AS sector`: Retrieves the SIC code, representing the company’s sector.
* `COUNT(DISTINCT t.tag) AS unique_tags`: Counts unique tags used by the company in its submissions.
* `SUM(n.value) AS total_assets`: Sums the value column in num, representing the company’s total assets.
* `JOIN num n ON s.adsh` = n.adsh: Links sub and num tables by adsh, matching submissions with numeric values.
* `JOIN tag t ON n.tag = t.tag AND n.version = t.version`: Connects num to tag by tag and version to access tag definitions.
* `WHERE s.filed BETWEEN '20240101' AND '20241231'`: Limits results to filings within 2024.
* `GROUP BY s.name, s.sic`: Groups results by company and sector.
* `ORDER BY total_assets DESC`: Orders by total_assets, descending, to show companies with the highest asset totals first.
* `LIMIT 5`: Displays the top 5 companies by total assets.

In [20]:
%%sql

SELECT 
    s.name AS company_name,
    s.sic AS sector,
    COUNT(DISTINCT t.tag) AS unique_tags,
    SUM(n.value) AS total_assets
FROM 
    sub s
JOIN 
    num n ON s.adsh = n.adsh
JOIN 
    tag t ON n.tag = t.tag AND n.version = t.version
WHERE 
    s.filed BETWEEN '20240101' AND '20241231'
GROUP BY 
    s.name, s.sic
ORDER BY 
    total_assets DESC
LIMIT 5;

 * postgresql://student@/DMA_G6
5 rows affected.


[('BANK OF CHILE', Decimal('6029'), 1066, Decimal('16048901542880200602.8597')),
 ('POSCO HOLDINGS INC.', Decimal('3312'), 687, Decimal('5155860969511875705.7398')),
 ('BANCO SANTANDER CHILE', Decimal('6029'), 1046, Decimal('3689022900181005786.4128')),
 ('KB FINANCIAL GROUP INC.', Decimal('6029'), 984, Decimal('148892842488236511.6123')),
 ('SHINHAN FINANCIAL GROUP CO LTD', Decimal('6021'), 1003, Decimal('148822933062439580.4438'))]